rasa_core/policies/form_policy.py

```diff
@@ -6,48 +6,130 @@
+        """Finds intents that shouldn't be validated.
+            Assumes that stories contain only unhappy paths"""
+
+        self.lookup = {}
 
-        pass
+        trackers_as_states, _ = self.featurizer.training_states_and_actions(
+                training_trackers, domain)
+
+        for states in trackers_as_states:
+            state = states[0]
+            if (any(ACTIVE_FORM_PREFIX in k and v > 0 for k, v in state.items()) and
+                    any(PREV_PREFIX + ACTION_LISTEN_NAME in k and v > 0
+                        for k, v in state.items())):
+                # by construction there is only one active form
+                form = [k[len(ACTIVE_FORM_PREFIX):]
+                        for k, v in state.items()
+                        if ACTIVE_FORM_PREFIX in k and v > 0][0]
+
+                feature_key = self._create_feature_key(states)
+                # even if there are two identical feature keys
+                # their form will be the same
+                self.lookup[feature_key] = form
+
+    @staticmethod
+    def _form_was_rejected(tracker):
+        # type: (DialogueStateTracker) -> bool
+        """Check whether previous call to the form was rejected"""
+        for event in reversed(tracker.applied_events()):
+            if (isinstance(event, ActionExecuted) and
+                    event.action_name == tracker.active_form):
+                return False
+            elif (isinstance(event, ActionExecutionRejected) and
+                    event.action_name == tracker.active_form):
+                return True
+
+        return False
+
+    @staticmethod
+    def _predicted_no_validation(tracker):
+        # type: (DialogueStateTracker) -> bool
+        """Check whether previous call to the form was rejected"""
+        for event in reversed(tracker.events):
+            if isinstance(event, ActionExecuted):
+                if event.action_name == ACTION_NO_FORM_VALIDATION_NAME:
+                    return True
+
+                break
+
+        return False
 
     def predict_action_probabilities(self, tracker, domain):
         # type: (DialogueStateTracker, Domain) -> List[float]
         """Predicts a form action if form is active"""
         result = [0.0] * domain.num_actions
 
         if tracker.active_form:
-            logger.debug("Form '{}' is active".format(tracker.active_form))
+            logger.debug("There is an active form '{}'"
+                         "".format(tracker.active_form))
             if tracker.latest_action_name == ACTION_LISTEN_NAME:
                 # predict form action after user utterance
-                idx = domain.index_for_action(tracker.active_form)
-                result[idx] = FORM_SCORE
+
+                if self._form_was_rejected(tracker):
+                    # since it is assumed that training stories contain
+                    # only unhappy paths, notify the form that
+                    # it should not be validated if predicted by other policy
```

Comment:
so if the user provides additional information when he gets asked if he wants to continue, this content is not used to fill the form, correct?

Reply:
no, if this information is not in training stories, then form will be predicted and validated